### PR TITLE
fix: fix a potential dos in the leaderboard command

### PR DESCRIPTION
**Pull request by thetayloredman**
**Does this Pull Request fix an issue? If so, please mention the issue.**
N/A

**Describe your changes.**
This PR will add a new method of managing ECTs (emote counter/trackers) which is DOS-proof when a user spams the leaderboard or API.

**Additional context**
Entry from private triage (maintainers only link): https://canary.discord.com/channels/771056568988860433/970479855579828245/972633654843043890

### CODE REVIEW

File: src/lib/structures/LexiECTManager.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        if (isOk) this._log.info(`ECTManager: ${db} list is ok`);
+        else {
+            this._log.error(`ECTManager: ${db} list is not ok... rebuilding.`);
+            this._rebuildReverse(db);
+        }
+    }
+
+    private _rebuildReverse(db: 'uwus' | 'owos' | 'tildes') {
+        const newArr: (string | never)[][] = [];
+        for (const [id, count] of this[db]) {
```
Comment (ilikestohack) [inline src/lib/structures/LexiECTManager.ts:90]: Kinda questioning the output of id and count maybe I'm missing what they actually are?

**Review by ilikestohack — DISMISSED** [review]
Looks alright I kinda got lost a bit but as long as it is tested it should be fine

File: src/lib/structures/LexiECTManager.ts (hunk 1 above)
Comment (thetayloredman) [reply src/lib/structures/LexiECTManager.ts:90]: id should be the user id
count is the number of entries

**Review by thetayloredman — COMMENTED** [review]
Forgot this lol

File: src/commands/leaderboard.ts [hunk 2]
```diff
@@ -61,15 +60,15 @@ export default class CatCommand extends LexiSlashCommand {
...
                 break;
             case 'owos':
                 buf.push('```adoc\n===== OWO LEADERBOARD =====');
-                sorted = client.emoteCounterTrackers.map((values, id) => [id, values.owos] as const).sort((a, b) => b[1] - a[1]);
+                sorted = client.emoteCounterTrackers.owosArray().sort((a, b) => b[1] - a[1]);
                 break;
             case 'tildes':
                 buf.push('```adoc\n===== TILDE LEADERBOARD =====');
-                sorted = client.emoteCounterTrackers.map((values, id) => [id, values.tildes] as const).sort((a, b) => b[1] - a[1]);
+                sorted = client.emoteCounterTrackers.tildesArray().sort((a, b) => b[1] - a[1]);
```
Comment (ilikestohack) [inline src/commands/leaderboard.ts:71]: But you are still sorting?

File: src/utils/web/index.ts [hunk 3]
```diff
@@ -35,47 +35,47 @@ export function start(client: Client, log: LexiLogger): void {
...
                 res.status(404).send('User not found');
                 return;
             }
             res.json({
                 uid: req.query.uid,
-                uwus: client.emoteCounterTrackers.get(req.query.uid as string, 'uwus')
+                uwus: client.emoteCounterTrackers.uwusForUser(req.query.uid as string)
             });
-        } else res.json(client.emoteCounterTrackers.map((values, id) => [id, values.uwus] as const).sort((a, b) => b[1] - a[1]));
+        } else res.json(client.emoteCounterTrackers.uwusArray().sort((a, b) => b[1] - a[1]));
```
Comment (ilikestohack) [inline src/utils/web/index.ts:46]: Still sorting?

File: src/utils/web/index.ts [hunk 4]
```diff
@@ -35,47 +35,47 @@ export function start(client: Client, log: LexiLogger): void {
...
                 res.status(404).send('User not found');
                 return;
             }
             res.json({
                 uid: req.query.uid,
-                uwus: client.emoteCounterTrackers.get(req.query.uid as string, 'owos')
+                uwus: client.emoteCounterTrackers.owosForUser(req.query.uid as string)
             });
-        } else res.json(client.emoteCounterTrackers.map((values, id) => [id, values.owos] as const).sort((a, b) => b[1] - a[1]));
+        } else res.json(client.emoteCounterTrackers.owosArray().sort((a, b) => b[1] - a[1]));
```
Comment (ilikestohack) [inline src/utils/web/index.ts:62]: Still sorting?

File: src/utils/web/index.ts [hunk 5]
```diff
@@ -35,47 +35,47 @@ export function start(client: Client, log: LexiLogger): void {
...
                 res.status(404).send('User not found');
                 return;
             }
             res.json({
                 uid: req.query.uid,
-                uwus: client.emoteCounterTrackers.get(req.query.uid as string, 'tildes')
+                uwus: client.emoteCounterTrackers.tildesForUser(req.query.uid as string)
             });
-        } else res.json(client.emoteCounterTrackers.map((values, id) => [id, values.tildes] as const).sort((a, b) => b[1] - a[1]));
+        } else res.json(client.emoteCounterTrackers.tildesArray().sort((a, b) => b[1] - a[1]));
```
Comment (ilikestohack) [inline src/utils/web/index.ts:78]: Still sorting?

**Review by ilikestohack — CHANGES_REQUESTED** [review]
Explain the use of the still existant sort function

File: src/commands/leaderboard.ts (hunk 2 above)
Comment (thetayloredman) [reply src/commands/leaderboard.ts:71]: The leaderboard component isn't done yet so this is just the needed polyfill right now

File: src/utils/web/index.ts (hunk 3 above)
Comment (thetayloredman) [reply src/utils/web/index.ts:46]: See previous comment

File: src/utils/web/index.ts (hunk 4 above)
Comment (thetayloredman) [reply src/utils/web/index.ts:62]: See previous comment

File: src/utils/web/index.ts (hunk 5 above)
Comment (thetayloredman) [reply src/utils/web/index.ts:78]: See previous comment

**Review by ilikestohack — APPROVED** [review]
Comments were explained

**Review by ilikestohack — DISMISSED** [review]
Looks good

**Comment by thetayloredman** [comment]
From the test so far
image

**Comment by thetayloredman** [comment]
Requires updates after #585 is merged

**Comment by 808-Dev** [comment]
Clarified JS analysis with submitter.

**Review by 808-Dev — DISMISSED** [review]
Reviewed and everything seems fine from here.

**Comment by thetayloredman** [comment]
@808-Dev @ilikestohack complete, review please?

**Review by ilikestohack — APPROVED** [review]
Didn't trace down everything but it looks alright

**Comment by thetayloredman** [comment]
> Didn't trace down everything but it looks alright
@ilikestohack 
i'll do some last-minute tests then i think we're green light

**Comment by thetayloredman** [comment]
WHOOPS

**Review by 808-Dev — APPROVED** [review]
As far as I know everything should be fine. Apologies for the wait

**Comment by thetayloredman** [comment]
I need an approving review to merge

**Review by ilikestohack — APPROVED** [review]
Seems fine to me but I didn't look in detail

**Pull request merged**